src/lib/spawn.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  ChildProcess,
+  spawn as cpSpawn,
+  SpawnOptions as cpSpawnOptions
+} from "child_process";
+
+import * as invariant from "invariant";
+
+export interface SpawnOptions extends cpSpawnOptions {
+  /**
+   * When true, unrefs the ChildProcess.
+   */
+  detached?: boolean;
+  /**
+   * populated when SpawnOtions.detached is true. Gives you a handle to the
+   * ChildProcess
+   */
+  child?: ChildProcess;
+}
+
+export class SpawnError extends Error {
```

Comment:
Note that writing a class extending an `Error` in TypeScript produces a class whose instances do not play well with the `instanceof` operator. I tested this by writing `class SpawnError extends Error {}` in the [TypeScript playground](http://www.typescriptlang.org/play/) and pasting the emitted code into a Node REPL:

```
kourge@gomibako.corp.eleveninc.com:~/projects/urbandoor/server (master)
  $ node
> class A {}
undefined
> class B extends A {}
undefined
> var a = new A()
undefined
> var b = new B()
undefined
> a instanceof A
true
> a instanceof B
false
> b instanceof B
true
> b instanceof A
true
> class SpawnError extends Error {}
undefined
> var e = new Error()
undefined
> var se = new SpawnError()
undefined
> e instanceof Error
true
> e instanceof SpawnError
false
> se instanceof SpawnError
true
> se instanceof Error
true
> kourge@gomibako.corp.eleveninc.com:~/projects/urbandoor/server (master)
  $ node
> var __extends = (this && this.__extends) || (function () {
...     var extendStatics = Object.setPrototypeOf ||
...         ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
...         function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
...     return function (d, b) {
.....         extendStatics(d, b);
.....         function __() { this.constructor = d; }
.....         d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
.....     };
... })();
undefined
> var SpawnError = /** @class */ (function (_super) {
...     __extends(SpawnError, _super);
...     function SpawnError() {
.....         return _super !== null && _super.apply(this, arguments) || this;
.....     }
...     return SpawnError;
... }(Error));
undefined
>
> var se = new SpawnError
undefined
> se instanceof SpawnError
false
> se instanceof Error
true
```

The [fix](https://github.com/UrbanDoor/server/blob/master/packages/data/parser/validator.ts#L7) is as obnoxious as it is slow:

```ts
class SpawnError extends Error {
  constructor() {
    super();
    Object.setPrototypeOf(this, SpawnError.prototype);
  }
}
```

If `instanceof` is not required to work, consider simply not including this single-line fix. See [these performance benchmarks](https://github.com/kourge/invokable/blob/master/README.md#performance) for how slow it is to [manually set the prototype of an object post-creation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf).

One possible workaround is exploiting the way constructors work in JavaScript:

```ts
class SpawnError extends Error {
  constructor() {
    super();
    return Object.create(SpawnError.prototype);
  }
}
```

It appears to do the right thing:

```
kourge@gomibako.corp.eleveninc.com:~/projects/urbandoor/server (master)
  $ node
> var __extends = (this && this.__extends) || (function () {
...     var extendStatics = Object.setPrototypeOf ||
...         ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
...         function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
...     return function (d, b) {
.....         extendStatics(d, b);
.....         function __() { this.constructor = d; }
.....         d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
.....     };
... })();
undefined
> var SpawnError = /** @class */ (function (_super) {
...     __extends(SpawnError, _super);
...     function SpawnError() {
.....         var _this = _super.call(this) || this;
.....         return Object.create(SpawnError.prototype);
.....     }
...     return SpawnError;
... }(Error));
undefined
> var se = new SpawnError()
undefined
> se instanceof SpawnError
true
> se instanceof Error
true
```

Reply:
I'll take a look at the output of tsc when targeting es2017. This got added to the language last year, so, it's supposed to work. Following the example of [semantic release](https://github.com/semantic-release/semantic-release/blob/caribou/docs/support/node-version.md), I'm targeting this at node 8 and later, so I *think* this'll be ok.

Reply:
yep, this is what tsc produced:

```js
class SpawnError extends Error {
    constructor() {
        super(...arguments);
        /**
         * exit code of the ChildProcess
         */
        this.code = 0;
        /**
         * stderr output of the ChildProcess
         */
        this.data = '';
    }
}
exports.SpawnError = SpawnError;
```

browsers and old node versions won't handle it well, but this particular library doesn't make sense in a browser and states that it needs node 8 or later.